sensing/autoware_cuda_pointcloud_preprocessor/src/cuda_downsample_filter/cuda_voxel_grid_downsample_filter.cu

```diff
@@ -0,0 +1,471 @@
+
+    // Fill the result
+    // Since referring pointer contents by `*min_max.first` and `*min_max.second`
+    // causes implicit data copy from device to host, download data using stream explicitly
+    CUDA_VGDF_ERROR_CHECK(cudaMemcpyAsync(
+      &min_field, thrust::raw_pointer_cast(min_max.first.get()), sizeof(float),
+      cudaMemcpyDeviceToHost, stream_));
+    CUDA_VGDF_ERROR_CHECK(cudaMemcpyAsync(
+      &max_field, thrust::raw_pointer_cast(min_max.second.get()), sizeof(float),
+      cudaMemcpyDeviceToHost, stream_));
+
+    return;
+  };  // auto & find_min_max
+
+  ThreeDim<float> min_coord{}, max_coord{};
+  find_min_max(0, min_coord.x, max_coord.x);
+  find_min_max(1, min_coord.y, max_coord.y);
+  find_min_max(2, min_coord.z, max_coord.z);
+
+  CUDA_VGDF_ERROR_CHECK(
+    cudaStreamSynchronize(stream_));  // Ensure the values on the host side ready to use
+
+  voxel_info_.min_coord.x = static_cast<int>(std::floor(min_coord.x / voxel_info_.voxel_size.x));
+  voxel_info_.max_coord.x = static_cast<int>(std::floor(max_coord.x / voxel_info_.voxel_size.x));
+  voxel_info_.min_coord.y = static_cast<int>(std::floor(min_coord.y / voxel_info_.voxel_size.y));
+  voxel_info_.max_coord.y = static_cast<int>(std::floor(max_coord.y / voxel_info_.voxel_size.y));
+  voxel_info_.min_coord.z = static_cast<int>(std::floor(min_coord.z / voxel_info_.voxel_size.z));
+  voxel_info_.max_coord.z = static_cast<int>(std::floor(max_coord.z / voxel_info_.voxel_size.z));
+}
+
+size_t CudaVoxelGridDownsampleFilter::searchValidVoxel(
+  const cuda_blackboard::CudaPointCloud2::ConstSharedPtr & points, size_t * voxel_index_buffer_dev,
+  size_t * point_index_buffer_dev, decltype(OutputPointType::return_type) * return_type_field_dev,
+  decltype(OutputPointType::channel) * channel_field_dev)
+{
+  // Update computation parameters on GPU constant memory
+  CUDA_VGDF_ERROR_CHECK(cudaMemcpyToSymbolAsync(
+    voxel_info_dev, &voxel_info_, sizeof(VoxelInfo), 0, cudaMemcpyDefault, stream_));
+
+  // calculate the number of voxel in each axis
+  ThreeDim<size_t> num_voxels = {
+    static_cast<size_t>(voxel_info_.max_coord.x - voxel_info_.min_coord.x) + 1,
+    static_cast<size_t>(voxel_info_.max_coord.y - voxel_info_.min_coord.y) + 1,
+    static_cast<size_t>(voxel_info_.max_coord.z - voxel_info_.min_coord.z) + 1,
+  };
+
+  // Define GPU kernel configuration
+  dim3 block_dim(512);
+  dim3 grid_dim((voxel_info_.num_input_points + block_dim.x - 1) / block_dim.x);
+
+  // calculate voxel index that each input point belong to
+  calculateVoxelIndexKernel<<<grid_dim, block_dim, 0, stream_>>>(
+    points->data.get(), num_voxels, voxel_index_buffer_dev, point_index_buffer_dev,
+    return_type_field_dev, channel_field_dev);
+
+  // get number of valid voxels
+  // Because thrust::unique_count returns the number of "consective" equal values,
+  // sorting is required beforehand to get all identical values
+  thrust::sort_by_key(
```

Comment:
Would there be any advantange of using  CUB here?
Sorting seems like the heaviest thing around and with CUB there is no attempt of allocation (which you have covered anyways) and you can save some bits since you know the ranges of the voxel coords (CUB allows to sort a subset of bits when using radix sort)

Reply:
I applied CUB sorting instead of thrust in 0e31c08f9. Thrust's sorting function uses CUB inside, but direct CUB API calls seem to reduce the number of kernels to be called. Thanks for the suggestion.

Reply:
@manato 
What about what I just mentioned about limiting the bits that are processed?
In the worst case (for a 0.3m grid) it uses less than half the iterations.
For a 3.0m grid it is even better

Reply:
@knzo25 
Thanks for pointing this out. I missed your old comment. I added a modification to set the most significant bit for radix sort in 0b9a1bb.